File: src/TopKFrequentElements.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.PriorityQueue;

public class TopKFrequentElements {
    public static void main(String[] args) {
        // test cases
        Object[] testCases = new Object[] {
                new Object[] { new int[] { 1, 1, 1, 2, 2, 3 }, 2, new int[] { 1, 2 } },
                new Object[] { new int[] { 1 }, 1, new int[] { 1 } }
        };

        for (Object testCase : testCases) {
            Object[] testCaseArray = (Object[]) testCase;
            int[] nums = (int[]) testCaseArray[0];
            int k = (int) testCaseArray[1];
            int[] expected = (int[]) testCaseArray[2];
            int[] actual = solution(nums, k);

            if (!Arrays.equals(expected, actual)) {
                throw new AssertionError(String.format("solution(%s, %s) == %s, expected %s",
                        Arrays.toString(nums), k, Arrays.toString(actual), Arrays.toString(expected)));
            }
        }
    }

    private static int[] solution(int[] nums, int k) {
        // count frequency of elements
        HashMap<Integer, Integer> counter = new HashMap<>();

        for (int num : nums) {
            counter.put(num, counter.getOrDefault(num, 0) + 1);
        }

        // create a max heap
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>((a, b) -> counter.get(b) - counter.get(a));

        for (int num : counter.keySet()) {
            maxHeap.add(num);
        }

        // write k most frequent elements to output
        int[] output = new int[k];

        for (int i = 0; i < k; i++) {
            output[i] = maxHeap.poll();
        }

        return output;

    }
}

File: python/topKFrequentElements.py
"""
347. Top K Frequent Elements

Given an integer array nums and an integer k, return the k most frequent elements. You may return the answer in any order.

Example 1:
Input: nums = [1,1,1,2,2,3], k = 2
Output: [1,2]

Example 2:
Input: nums = [1], k = 1
Output: [1]


Constraints:
1 <= nums.length <= 10**5
-10**4 <= nums[i] <= 10**4
k is in the range [1, the number of unique elements in the array].
It is guaranteed that the answer is unique.
"""

import heapq
from typing import List

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # easy solution -> heap  -> O(n*logn) time
        # counter
        frequency = {}
        for num in nums:
            frequency[num] = 1 + frequency.get(num, 0)

        priority_queue = []
        for num, freq in frequency.items():
            # map value: -frequency: num
            # note: heapq is a minheap, so convert to negative
            freq *= -1
            heapq.heappush(priority_queue, (freq, num))

        # append the top-k items in priority_queue to result
        result = []
        while priority_queue:
            result.append(heapq.heappop(priority_queue)[1])
            if len(result) ==  k:
                return result

        # better solution -> hashMap -> O(n) time
        counter = {}
        freq = [[] for _ in range(len(nums)+1)]

        for num in nums:
            counter[num] = 1 + counter.get(num, 0)

        for key, val in counter.items():
            freq[val].append(key)

        res = []
        for i in range(len(nums), 0, -1):
            for num in freq[i]:
                res.append(num)

                if len(res) == k:
                    return res

if __name__ == "__main__":
    INPUTS = (
        # nums  k   expected
        ([1,1,1,2,2,3], 2,  [1,2]),
        ([1],           1,  [1]),
    )

    for nums, k, expected in INPUTS:
        actual = Solution().topKFrequent(nums, k)
        assert actual == expected, f"{actual=}, {expected=}, {nums=}, {k=}"
